conans/client/build/cmake_toolchain_build_helper.py

```diff
@@ -47,198 +43,103 @@ class CMakeToolchainBuildHelper(BaseCMakeBuildHelper):
-            verbose = self._definitions[verbose_definition_name]
-            return get_bool_from_text(str(verbose))
-        except KeyError:
-            return False
-
-    @verbose.setter
-    def verbose(self, value):
-        self._definitions.update(verbose_definition(value))
-
-    def _get_dirs(self, source_folder, build_folder, cache_build_folder):
-
-        def get_dir(folder, origin):
-            if folder:
-                return os.path.join(origin, folder)
-            return origin
-
-        build_ret = get_dir(build_folder, self._conanfile.build_folder)
-        source_ret = get_dir(source_folder, self._conanfile.source_folder)
-
-        if self._conanfile.in_local_cache and cache_build_folder:
-            build_ret = get_dir(cache_build_folder, self._conanfile.build_folder)
-
-        return source_ret, build_ret
-
-    def configure(self, args=None, defs=None, source_dir=None, build_dir=None,
-                  source_folder=None, build_folder=None, cache_build_folder=None,
-                  pkg_config_paths=None):
-        assert args is None, "All the 'args' should be provided in the toolchain"
-        assert defs is None, "All the 'defs' should be provided in the toolchain"
-        assert source_dir is None, "'source_dir' is deprecated. Use 'source_folder'"
-        assert build_dir is None, "'build_dir' is deprecated. Use 'source_folder'"
-        assert pkg_config_paths is None, "'pkg_config_paths' should be provided in the toolchain"  # TODO: environment?
-
+    def configure(self, source_folder=None, build_folder=None, build_type=None):
+        # TODO: environment?
         if not self._conanfile.should_configure:
             return
 
-        defs = {"CMAKE_TOOLCHAIN_FILE": CMakeToolchain.filename}
-        configure_command = self._configure_command(defs, source_folder, build_folder,
-                                                    cache_build_folder)
+        sf = self._conanfile.source_folder
+        if source_folder:
+            sf = os.path.join(self._conanfile.source_folder, source_folder)
 
-        if self._is_windows_mingw:
-            with tools.remove_from_path("sh"):
-                self._conanfile.run(configure_command)
-        else:
-            self._conanfile.run(configure_command)
+        bf = self._conanfile.build_folder
+        if build_folder or self._build_folder:
+            bf = os.path.join(self._conanfile.build_folder, build_folder or self._build_folder)
+
+        bt = build_type or self._build_type or self._conanfile.settings.get_safe("build_type")
 
-    def _configure_command(self, defs, source_folder=None, build_folder=None,
-                           cache_build_folder=None):
-        source_dir, self.build_dir = self._get_dirs(source_folder, build_folder, cache_build_folder)
-        mkdir(self.build_dir)
+        defs = {"CMAKE_TOOLCHAIN_FILE": CMakeToolchain.filename}
+        if bt and not self._is_multiconfiguration:
+            defs["CMAKE_BUILD_TYPE"] = bt
```

Comment:
This is already assigned in the toolchain, remove

Reply:
But the toolchain cannot assign the build type for multi-config like Visual Studio. If the toolchain is for the user/developer flow, this will not work. Or the approach is:

- In single-config: have 1 toolchain per configuration that defines the build_type
- In multi-config: have 1 toolchain that does not define the build_type, the build type is passed later by the build helper

Reply:
This definition is not applied for multi-config generators. For a single-config generator the build_type is applied by the toolchain.